samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,979 @@
+		find_headers_frame_stream_id(&frame);
+
+	printf("||stream id %d||\n",
+		ctx_client->streams[ctx_client->stream_count].stream_id);
+
+	if (!has_upgrade_header) {
+		ctx_client->streams[0].stream_id = 1;
+		unsigned char response_headers_frame[16];
+
+		ctx_client->streams[ctx_client->stream_count].stream_state =
+			OPEN_STATE;
+
+		generate_response_headers_frame(response_headers_frame,
+			ctx_client->streams[0].stream_id);
+
+		size_t ret = sendall(ctx_client->client_fd,
+			response_headers_frame, sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+
+		ctx_client->stream_count = 1;
+	}
+
+	if (settings_end_headers_flag(ctx_client->buffer[4])) {
+		unsigned char response_headers_frame[16];
+
+		generate_response_headers_frame(response_headers_frame,
+			ctx_client->streams[ctx_client->stream_count]
+				.stream_id);
+
+		ctx_client->stream_count++;
+
+		size_t ret = sendall(ctx_client->client_fd,
+			response_headers_frame, sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+	} else {
+		/* We expect Continuation frame */
+		ctx_client->server_state = HTTP2_FRAME_HEADER_STATE;
+		return 0;
+	}
+
+	unsigned int frame_size = determine_frame_size(ctx_client->buffer);
+
+	ctx_client->offset -= frame_size;
+	memmove(ctx_client->buffer, ctx_client->buffer + frame_size,
+		ctx_client->offset);
+
+	if (ctx_client->offset == 0) {
+		if (ctx_client->streams[ctx_client->stream_count]
+				.stream_state == IDLE_STATE) {
+			handle_http2_open_state(ctx_client);
+		} else {
+			ctx_client->server_state = HTTP2_FRAME_HEADER_STATE;
+		}
+	} else {
+		ctx_client->server_state =
+			determine_server_state(ctx_client->buffer);
+	}
+
+	return 0;
+}
+
+int handle_http2_open_state(struct http2_client_ctx *ctx_client)
+{
+	printf("OPEN_STATE\n");
+
+	parse_http2_frame(ctx_client->buffer, ctx_client->offset, &frame);
+	print_http2_frames(&frame);
+
+	size_t content_size = sizeof(content);
+
+	for (int i = 0; i < ctx_client->stream_count; i++) {
+		send_data(ctx_client->client_fd, content, content_size, 0x00,
+			0x01, ctx_client->streams[i].stream_id);
+	}
+
+	ctx_client->streams[0].stream_state = -1;
+
+	if (has_upgrade_header)
+		if (ctx_client->offset == 0) {
+			ctx_client->server_state = HTTP2_FRAME_HEADER_STATE;
+		} else {
+			ctx_client->server_state =
+				determine_server_state(ctx_client->buffer);
+		}
+	else
+		ctx_client->server_state = HTTP2_FRAME_GOAWAY_STATE;
+
+	has_upgrade_header = true;
+
+	return 0;
+}
+
+int handle_http2_frame_headers(struct http2_client_ctx *ctx_client)
+{
+	printf("HTTP2_FRAME_HEADERS\n");
+
+	ctx_client->streams[0].stream_state =
+		determine_stream_state(ctx_client->buffer);
+	int ret = -EINVAL;
+
+	switch (ctx_client->streams[0].stream_state) {
+	case IDLE_STATE:
+		ret = handle_http2_idle_state(ctx_client);
+		break;
+	case OPEN_STATE:
+		ret = handle_http2_open_state(ctx_client);
+		break;
```

Comment:
I don't really think we should need separate handlers for IDLE/OPEN stream state. First off, as soon we allocate the stream it should be marked as OPEN, so there shouldn't really be the case we handle IDLE state. We should focus here on parsing header fields, one by one, and on each header you parse you should decrement `current_frame->length` by the length you parsed. And when `current_frame->length` reaches 0, it means that we're done with the current frame and we should expect a new one (so state transition to HTTP2_FRAME_HEADER_STATE, again this may involve some action, like sending the response if we received STREAM_END).

For the headers that are essential for serving the request (like the resource URI the request refers to) you should have a variable within `struct stream_context`, so that once we're done with request parsing (STREAM_END received) you know which resource to fetch. There are probably more, this is just an example.

And finally, if it turns out, that `current_frame->length` is still >0, but there's no more data in the buffer to parse, it means that we did not receive entire HEADERS frame in a single TCP packet (which can happen with TCP!). In that case, return -EAGAIN, and let `poll()`/`recv()` to fill the buffer with new data. When we're back in `handle_http_request()` it should be clear that:
* we are in HTTP2_FRAME_HEADERS_STATE
* we know how much data is left to parse in current frame (`current_frame->length`),
so we should proceed with parsing individual headers.